rfcs/react-components/convergence/component-transitions-on-mount-or-unmount.md

```diff
@@ -0,0 +1,125 @@
+# RFC: Component CSS Transitions on mount/unmount
+
+<!--
+An RFC can be anything. A question, a suggestion, a plan. The purpose of this template is to give some structure to help folks write successful RFCs. However, don't feel constrained by this template; use your best judgement.
+
+Tips for writing a successful RFC:
+
+- Simple plain words that make your point, fancy words obfuscate
+- Try to stay concise, but don't gloss over important details
+- Try to write a neutral problem statement, not one that motivates your desired solution
+- Remember, "Writing is thinking". It's natural to realize new ideas while writing your proposal
+-->
+
+---
+
+@marcosmoura
+
+## Summary
+
+This RFC describes how CSS transitions in Fluent UI components can be applied.
+
+## Background
+
+Currently it is not possible to add CSS transition to components on mount/unmount. A usage of this pattern would be transitions to components that has a open/close toggle, such as Drawer, Dialog and more.
+
+## Problem statement
+
+In order to display that a content is showing or hiding from screen, CSS transitions should be applied. Due to React's way of rendering, there is no builtin support for this, therefore we need to create a solution that can provide with the status of component on screen.
+
+## Detailed Design or Proposal
+
+### A `useAnimationState` hook based solution
+
+To differentiate the animation state of a component, a hook could be created and be part of the react-utilities. Note that every time a status change (from not visible to entering for example) the component will rerender. A draft implementation of the hook: [useAnimationState](https://github.com/marcosmoura/fluentui/blob/marcosmoura/feat/drawer-component/packages/react-components/react-drawer/src/components/Drawer/useAnimationState.ts)
+
+Usage in the component:
+
+```ts
+import * as React from 'react';
+import { getNativeElementProps, useAnimationState } from '@fluentui/react-utilities';
```

Comment:
If we are going to use a hook based approach, it would be worth checking out if we can improve how to handle transitions of collections of components. Here's an example

```tsx
// each `Item` components needs to be animated on enter and exit
// If I remove an element from the `items` array how is it possible to handle unmount
return (
  <div>
   {items.map(item => <Item />)}
  </div>
)
```

The react-transition-group package does some hacky things with mapping children if we go for a hooks approach we could be smarter. Here's the `TransitionGroup` code that does children collection handling for reference https://github.com/reactjs/react-transition-group/blob/2989b5b87b4b4d1001f21c8efa503049ffb4fe8d/src/TransitionGroup.js#L70-L87

Reply:
I've experimented with this already. It is currently possible to add this behavior using a hook-based solution, but the implementation it is quite different from React Transition Group. I think this deserve a separate RFC for this matter.